RFC: Bring back apropos

mbauman opened the pull request:
This adds a very simple documentation search in order to bring back apropos:

``` jl
julia> apropos("pearson")
cov
cor

julia> apropos("combine")
Base.Docs.catdoc
muladd
ldltfact
sparsevec
sparse
Base.include_from_node1
cholfact
graphemes
Base.Profile.print
bfft

julia> import DataFrames

julia> apropos(r"\bNA\b")
DataFrames.complete_cases
DataFrames.nonunique
DataFrames.complete_cases!
StatsBase.describe
DataFrames.AbstractDataFrame
join
```

I've also updated the display of functions to include module names when they're not exported from Base.  This is important in cases like `Profile.print`, where it would currently just write `print`.

One downside here is that it cannot search across multiple markdown elements.  But this is the simplest solution that I saw to get something basic working… and it works quite well in most simple cases.  On the plus side, this way is extremely fast.

mbauman committed:
Bring back apropos

This adds a very simple documentation search in order to bring back apropos:

```jl
julia> apropos("pearson")
cov
cor

julia> apropos("combine")
Base.Docs: catdoc
graphemes
bfft
muladd
sparse
ldltfact
print
include_from_node1
cholfact
sparsevec
```

mbauman commented on doc/stdlib/base.rst:
```diff
@@ -156,6 +156,10 @@ Getting Around
+.. function:: apropos(string)
+
+   .. Docstring generated from Julia source
+
```
Why can't `genstdlib.jl` find apropos?  It's exported from Base.Docs, Base is `using .Docs`, and it's exported from Base.

yuyichao replied:
You don't need the `.. Docstring generated from Julia source` line, it is added automatically. What's the warning genstdlib gives about missing apropos (which module)? Are you sure you are running it with the version of julia compiled with the doc?

yuyichao replied:
Also I think we are only looking in the meta for `Base` so documenting it inline in a different module may not work.

yuyichao replied:
@mbauman Can you check if the [yyc/genstdlib-module](https://github.com/JuliaLang/julia/tree/yyc/genstdlib-module) branch fixes it?

mbauman replied:
Did you push that branch?  I don't see it.  The error is simply:

```
$ ./julia doc/genstdlib.jl
Collect 1481 Docs

Converting stdlib/

KeyError(apropos)
WARNING: Mod Base apropos
INFO: no docs for apropos(string) in Base
```

I figured it was probably the submodule thing, but hadn't dug into it yet.  It'd be great to be able to put documentation in submodules (like Dates, FFTs, Docs, etc).

ScottPJones replied:
Totally agree, my take is that we should be able to document everything, whether it is exported,  meant to be "unexported but public" (like things in Pkg) or even "unexported meant to be private", no matter where it is in modules.

yuyichao replied:
@mbauman That branch was already merged and deleted. Can you try rebasing on the current master.

mbauman committed:
Show module names of functions not in stdlib

Also refactor TypeName display to use the same machinery. Make it a bit more robust for names defined in submodules of Base or Core.

mbauman commented:
Hah!  You're too fast.  That's wonderful.  Fixed. :+1:

yuyichao commented:
Just want to test if the fix works ;-p .

mbauman committed:
Add simple tests

jakebolewski commented:
![Simpsons](http://cdn.meme.am/instances/500x/54833458.jpg)

mbauman commented:
Whoa whoa whoa.  Gah!  That was entirely too easy:

```
$ history
…
13933  git push origin master mb/apropos -f
…
```

Dammit.  Sorry.  :(

mbauman commented:
I have a local copy that seems to be more up-to-date:

```
* 7084ea7 2015-08-25 | Remove deprecated test_fails [Iain Dunning]
* 49ab228 2015-09-04 | Check try bindings with other modules [Yichao Yu]
* f9e08a5 2015-09-04 | Add missing doc for checkbounds [Yichao Yu]
* 3737c0a 2015-09-04 | Actually count exported functions [Yichao Yu]
* ac1e375 2015-09-04 | Look for docs in modules other than Base. [Yichao Yu]
* 5c79ba1 2015-09-04 | fix jl_recache_type in incremental deserializer [Jameson Nash]
* ee90995 2015-09-04 | add Tuple{Vararg{Char}} to  Union, closes #12918 [Jake Bolewski]
* 1fcdca3 2015-08-31 | make Cstring more usable as a return value (see #10799) [Steven G. Johnson]
* fb65488 2015-08-31 | Optimize several Complex{Bool} operations [Erik Schnetter]
```

Should I push that?

KristofferC commented:
Time to activate https://help.github.com/enterprise/11.10.340/admin/articles/blocking-force-pushes-to-a-repository?  It's an easy mistake to do.

jakebolewski commented:
I need to go but you need to dig out the last good commit SHA (if you have it lying around in another repo, or you can retrieve it from the reflog) and do a hard reset of your history before pushing the update.

mbauman commented:
Hrm. Ok. I'll see what I can do.

jiahao commented:
Here is what the [GitHub API](https://api.github.com/repos/JuliaLang/julia/events) recorded:

```
{
    "id": "3121339788",
    "type": "PushEvent",
    "actor": {
      "id": 154641,
      "login": "mbauman",
      "gravatar_id": "",
      "url": "https://api.github.com/users/mbauman",
      "avatar_url": "https://avatars.githubusercontent.com/u/154641?"
    },
    "repo": {
      "id": 1644196,
      "name": "JuliaLang/julia",
      "url": "https://api.github.com/repos/JuliaLang/julia"
    },
    "payload": {
      "push_id": 782193776,
      "size": 0,
      "distinct_size": 0,
      "ref": "refs/heads/master",
      "head": "b122bc7d31eb0c25e9e88c819b9f1d07a803ffa9",
      "before": "b3a1be57e40330db68734778c23d79726c2da0b9",
      "commits": [

      ]
    },
    "public": true,
    "created_at": "2015-09-05T15:03:59Z",
    "org": {
      "id": 743164,
      "login": "JuliaLang",
      "gravatar_id": "",
      "url": "https://api.github.com/orgs/JuliaLang",
      "avatar_url": "https://avatars.githubusercontent.com/u/743164?"
    }
  },
  ```

Looks like you'll want b3a1be57e40330db68734778c23d79726c2da0b9

jiahao commented:
cc @ihnorton that's your commit. I think we'll be fine if you push your local master HEAD

mbauman commented:
Aha, found it in my reflog.  Now I'm at

```
$ git status
HEAD detached at b3a1be5
nothing to commit, working directory clean
```

Working on getting it back to my local `master` with a clean history now.

KristofferC commented:
Would `git reset b3a1be5 --hard` work?

ihnorton commented:
@mbauman I'll go ahead and force push it.

mbauman commented:
Ah, wonderful.  Thanks Isaiah.  Again, terribly sorry for the snafu.

Keno commented:
Sounds like GitHub will get protected branches soon, which would prevent this.

KristofferC commented:
See https://github.com/blog/2051-protected-branches-and-required-status-checks

ihnorton commented:
All set now. Two weeks time-out from commit bit for @mbauman! :palm_tree:

I have my git configured to always require passwords for julialang/julia. Not foolproof, but does require at least a second glance.

Keno commented:
Yes, but it's not enabled yet for this repository, I already checked.

jiahao commented:
No worries @mbauman. Thankfully git+GitHub save enough data to reconstruct what happened. (And it could have been much worse: https://news.ycombinator.com/item?id=6713742)

mbauman commented:
I think I'll be trying something like: https://gist.github.com/pixelhandler/5718585 locally.  Looks like a nice addition.

mbauman committed:
Change test for UniformScaling name

nalimilan commented:
FWIW, I have switched to typing `git push origin +nl/mybranch` instead of `git push --force`. That saved me a few mistakes.

jakebolewski commented:
I liked this function as well.  Maybe an expanded future version could print out the docstrings with the match highlighted.

mbauman commented:
Awesome!  Thanks!

jakebolewski commented:
Maybe you can elaborate on "multiple markdown elements", I tried `apropos("foo")` and it was searching correctly searching through the Markdown code blocks.

mbauman commented:
Yes, it's just text that spans across multiple Markdown styles that will cause trouble.  Eg., "foo bar" won't be found in `*foo* bar`.

jakebolewski commented:
You should file an "up for grabs" issue for this.
